crates/red_knot_python_semantic/resources/mdtest/attributes.md

```diff
@@ -277,6 +316,39 @@ reveal_type(C.variable_with_class_default1)  # revealed: str
 reveal_type(c_instance.variable_with_class_default1)  # revealed: str
 ```
 
+### Inheritance of class/instance attributes
+
+#### Instance variable defined in a base class
+
+```py
+class Base:
+    declared_in_body: int | None = 1
+
+    can_not_be_redeclared: str | None = None
+
+    def __init__(self) -> None:
+        self.defined_in_init: str | None = "value in base"
+
+class Intermediate(Base):
+    # TODO: Mypy does not report an error here, but pyright does:
+    # "… overrides symbol of same name in class "Base". Variable is mutable so its type is invariant"
+    # We should introduce a diagnostic for this. Whether or not that should be enabled by default can
+    # still be discussed.
+    can_not_be_redeclared: str = "a"
```

Comment:
I think there are two possible rules here: "subclass can not re-declare, period", or "subclass can only re-declare with an equivalent type, because mutability implies invariance". The name `can_not_be_redeclared` here suggests the former, but pyright implements the latter. There's no unsoundness to redeclare with the same type on a subclass, and pyright allows it without error. Perhaps we could be a little clearer about that here? Even show an example of both?

Reply:
Thanks. https://github.com/astral-sh/ruff/pull/15826